### PR TITLE
Stop cloning events when adding to state

### PR DESCRIPTION
**Pull request by dbkr**
As comment hopefully explains.

On my test account:
Before: 394657 MatrixEvents, 53MB shallow size
After: 198863 MatrixEvents, 27MB shallow size

### CODE REVIEW

File: src/models/event-timeline.js [hunk 1]
```diff
@@ -88,19 +86,19 @@ EventTimeline.prototype.initialiseState = function(stateEvents) {
...
-
-    this._startState.setStateEvents(oldStateEvents);
+    // We previously deep copied events here and used different copies in
+    // the oldState and state events: this decision seems to date back
+    // quite a way and was apparently made to fix a bug where modifications
+    // made to the start state leaked through to the end state.
+    // This really shouldn't be possible though: the events themselves should
+    // not change. Duplicating the events uses a lot of extra memory,
+    // so we now no longer do it. To assert that they really do never change,
+    // freeze them! Note that we can't do this for events in general:
```
Comment (richvdh) [inline src/models/event-timeline.js]: (why not?)

File: src/models/event-timeline.js [hunk 2]
```diff
@@ -88,19 +86,19 @@ EventTimeline.prototype.initialiseState = function(stateEvents) {
...
+    for (const e of stateEvents) {
+        Object.freeze(e);
```
Comment (richvdh) [inline src/models/event-timeline.js]: do we need to do some babel foo to make this work on node?

Comment (dbkr) [reply src/models/event-timeline.js]: Remarkably no: Object.freeze has been supported in node basically forever: http://node.green/#ES2015-misc-Object-static-methods-accept-primitives-Object-freeze

File: src/models/event-timeline.js (hunk 1 above)
Comment (dbkr) [reply src/models/event-timeline.js]: have elaborated

**Pull request merged**